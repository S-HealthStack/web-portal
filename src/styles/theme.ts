import { css, DefaultTheme } from 'styled-components';
import { px } from './utils';

const colors = {
  // gray gradations
  black: '#000000',
  black40: 'rgba(0, 0, 0, 0.4)',
  black08: 'rgba(0, 0, 0, 0.08)',

  // light mode
  surface: '#FFFFFF',
  background: '#F7F8FA',
  onPrimary: '#FFFFFF',
  backgroundOnPrimary: '#FFFFFF',
  backgroundSurface: '#FFFFFF',
  backgroundOnSurface: '#2D2D2D',
  primary: '#4475E3',
  primaryLight: '#ECF1FC',
  primaryHovered: '#2F65E0',
  primary05: '#F6F8FE',
  primary10: '#ECF1FC',
  primary30: '#C7D6F7',
  primaryDisabled: '#B3C6F1',
  primaryLightPressed: '#E3EAFB',
  primaryLightFocused: '#C7D6F7',
  textPrimary: '#474747',
  primaryWhite: '#FFFFFF',
  disabled: '#DBDBDB',
  disabled20: 'rgba(219, 219, 219, 0.2)',
  onDisabled: '#A1A1A1',

  textPrimaryBlue: '#4475E3',
  textPrimaryDark: '#474747',
  textSecondaryGray: '#808080',
  textDisabled: '#838383',
  onSurface: '#2D2D2D',
  secondarySkyblue: '#00B0D7', // replace
  backgoundLight: '#F0F0F0',
  primaryBlueHovered: '#2F65E0',
  primaryBluePressed: '#2057D5',

  statusSuccess: '#2DC008',
  statusSuccess10: '#DCF2EA',
  statusSuccessText: '#00825D',

  statusError: '#FF3333',
  statusError10: '#FFEBEB',
  statusErrorText: '#D14343',

  statusWarning: '#F5BF00',
  statusWarning10: '#FFF7D0',
  statusWarningText: '#A36400',

  secondaryViolet: '#944ED7',
  secondaryVioletHovered: '#6A3799',
  secondaryViolet10: '#F4EDFB',

  secondaryTangerine: '#F39C18',
  secondaryTangerineHovered: '#CC8314',
  secondaryTangerine10: '#FEF5E8',

  secondarySkyBlue: '#00B0D7',
  secondarySkyBlueHovered: '#0092B3',
  secondarySkyBlue10: '#E5F7FB',

  secondaryGreen: '#1AD598',
  secondaryGreenHovered: '#16B37F',
  secondaryGreen10: '#E8FBF5',

  secondaryRed: '#FA5F4F',
  secondaryRedHovered: '#CC4E40',
  secondaryRed10: '#FFEFED',
};

export type SpecColorType = keyof typeof colors;

const weightToCssValue = {
  regular: 400,
  upperregular: 450,
  medium: 500,
  semibold: 600,
  bold: 700,
};

const fontSpec = (
  size: number,
  lineHeight: number | string,
  weight: keyof typeof weightToCssValue,
  fontFamily: string
) => css`
  font-family: ${fontFamily}, sans-serif;
  font-size: ${px(size)};
  line-height: ${typeof lineHeight === 'string' ? lineHeight : px(lineHeight)};
  font-weight: ${weightToCssValue[weight]};
  font-style: normal;
`;

const typography = {
  // Inter typeface
  labelRegular: fontSpec(10, 15, 'regular', 'Inter'),
  labelSemibold: fontSpec(10, 15, 'semibold', 'Inter'),

  bodyXSmallRegular: fontSpec(12, 18, 'regular', 'Inter'),
  bodyXSmallSemibold: fontSpec(12, 18, 'semibold', 'Inter'),
  bodySmallRegular: fontSpec(14, 21, 'regular', 'Inter'),
  bodySmallSemibold: fontSpec(14, 21, 'semibold', 'Inter'),
  bodyMediumRegular: fontSpec(16, 24, 'regular', 'Inter'),
  bodyMediumSemibold: fontSpec(16, 24, 'semibold', 'Inter'),
  bodyLargeRegular: fontSpec(18, 26, 'regular', 'Inter'),
  bodyLargeSemibold: fontSpec(18, 26, 'semibold', 'Inter'),

  headingXSmall: fontSpec(16, 21, 'semibold', 'Inter'),
  headingSmall: fontSpec(18, 23, 'semibold', 'Inter'),
  headingXMedium: fontSpec(18, 27, 'semibold', 'Inter'),
  headingXMediumRegular: fontSpec(18, 27, 'regular', 'Inter'),
  headingMedium: fontSpec(24, 32.2, 'semibold', 'Inter'),
  headingLargeSemibold: fontSpec(32, 42, 'semibold', 'Inter'),
  headingXLargeSemibold: fontSpec(48, 58, 'semibold', 'Inter'),
  headingXXLargeSemibold: fontSpec(60, 72, 'semibold', 'Inter'),

  // Fira Code typeface
  query14: fontSpec(14, '130%', 'upperregular', 'Fira Code'),
  query12: fontSpec(12, '150%', 'upperregular', 'Fira Code'),
};

const boxShadow = {
  card: `0 0 ${px(2)} ${px(0)} rgba(0,0,0,0.15)`,
  avatar: `0 ${px(8)} ${px(10)} ${px(2)} 0 rgba(0,0,0,0.12)`,
  studyAvatar: `0 ${px(15)} ${px(20)} rgba(180,180,180,0.3)`,
  previewScreen: `0 ${px(2)} ${px(4)} rgba(71,71,71,0.25)`,
};

const animation = {
  defaultTiming: 'cubic-bezier(0.45, 0.05, 0, 1)',
  rippleTiming: 'cubic-bezier(0.4, 0, 0.2, 1)',
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: typeof colors;
    typography: typeof typography;
    boxShadow: typeof boxShadow;
    animation: typeof animation;
  }
}

const theme: DefaultTheme = {
  colors,
  typography,
  boxShadow,
  animation,
};

export default theme;
