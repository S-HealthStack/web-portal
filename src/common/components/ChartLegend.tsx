import React, { FC } from 'react';
import styled from 'styled-components';
import _capitalize from 'lodash/capitalize';

import { SpecColorType } from 'src/styles/theme';
import { px, colors, typography } from 'src/styles';
import Radio from './Radio';

const Container = styled.div<{ mode: ChartLegendMode }>`
  display: flex;
  justify-content: ${({ mode }) => (mode === 'left' ? 'start' : 'space-between')};
`;

const ItemWrapper = styled.div<{ $isLast?: boolean; $opacity?: number; mode: ChartLegendMode }>`
  margin-right: ${({ mode }) => (mode === 'left' ? px(10) : 0)};

  circle {
    opacity: ${({ $opacity }) => $opacity || 1};
  }
`;

const RadioLabel = styled.div<{ $canToggle: boolean; $reverse: boolean }>`
  ${typography.bodySmallRegular};
  color: ${colors.onSurface};
  display: flex;
  align-items: center;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  width: 100%;
  -webkit-line-clamp: 2; /* number of lines to show */
  line-clamp: 2;
  -webkit-box-orient: vertical;
  white-space: pre-line;
`;

export type ChartLegendItem = {
  name: string;
  color: SpecColorType;
  checked: boolean;
  ignoreAsLast?: boolean;
  opacity?: number;
};

export type ChartLegendMode = 'left' | 'space-between';

export type ChartLegendProps = {
  items: ChartLegendItem[];
  canToggle: boolean;
  mode?: ChartLegendMode;
  onChange?: (index: number) => void;
};

const ChartLegend: FC<ChartLegendProps> = ({
  items,
  canToggle,
  mode = 'left',
  onChange,
  ...restProps
}) => {
  const isOneItemChecked = items
    ? items.filter(({ ignoreAsLast, checked }) => !ignoreAsLast && checked).length === 1
    : false;

  return (
    <Container mode={mode} {...restProps}>
      {items?.map((item: ChartLegendItem, index: number) => {
        const reverse = mode === 'space-between' && index !== 0;

        return (
          <ItemWrapper
            $isLast={index === items.length - 1}
            key={item.name}
            $opacity={item.opacity}
            mode={mode}
          >
            <Radio
              data-testid={item.name}
              key={item.name}
              kind={canToggle ? 'radio' : 'filled'}
              color={item.color}
              checked={!canToggle || item.checked}
              onChange={() => canToggle && onChange && onChange(index)}
              disabled={canToggle && isOneItemChecked && item.checked && !item.ignoreAsLast}
              reverse={reverse}
              readOnly={!canToggle}
              isLegend
            >
              <RadioLabel $canToggle={canToggle} $reverse={reverse}>
                {_capitalize(item.name)}
              </RadioLabel>
            </Radio>
          </ItemWrapper>
        );
      })}
    </Container>
  );
};

export default ChartLegend;
