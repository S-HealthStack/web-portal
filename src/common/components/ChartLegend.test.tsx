import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { getByText, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import theme from 'src/styles/theme';
import { ThemeProvider } from 'styled-components/';
import ChartLegend, { ChartLegendItem, ChartLegendProps } from './ChartLegend';

describe('ChartLegend', () => {
  const items: Array<ChartLegendItem> = [
    {
      name: 'male',
      color: 'primary',
      checked: false,
    },
    {
      name: 'female',
      color: 'secondaryGreen',
      checked: false,
    },
  ];

  it('test chart legend render', async () => {
    const onChange = jest.fn();

    const { baseElement, getByTestId, queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <ChartLegend items={items} data-testid="chart-legend" canToggle onChange={onChange} />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const chartLegend = getByTestId('chart-legend');

    expect(chartLegend).toBeInTheDocument();
    expect(getByText(baseElement, 'Male')).toBeInTheDocument();
    expect(getByText(baseElement, 'Female')).toBeInTheDocument();

    const radioMale = queryByTestId('male') as Element;
    const radioFemale = queryByTestId('female') as Element;

    expect(radioMale).not.toBeChecked();
    expect(radioFemale).not.toBeChecked();

    await userEvent.click(radioMale);

    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it('test chart legend if canToggle=false', async () => {
    const { baseElement, getByTestId, queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <ChartLegend items={items} data-testid="chart-legend" canToggle={false} />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const chartLegend = getByTestId('chart-legend');
    const radioMale = queryByTestId('male') as Element;
    const radioFemale = queryByTestId('female') as Element;

    expect(chartLegend).toBeInTheDocument();

    expect(radioMale).toBeChecked();
    expect(radioFemale).toBeChecked();

    await userEvent.click(radioMale);

    await waitFor(() => {
      expect(radioMale).toBeChecked();
      expect(radioFemale).toBeChecked();
    });
  });

  it('[NEGATIVE] should render chart legend with empty props', async () => {
    const onChange = jest.fn();

    const { baseElement, getByTestId } = render(
      <ThemeProvider theme={theme}>
        <ChartLegend items={[]} data-testid="chart-legend" canToggle onChange={onChange} />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const chartLegend = getByTestId('chart-legend');

    expect(chartLegend).toBeInTheDocument();
  });

  it('[NEGATIVE] should render chart legend without props', async () => {
    const fakeProps = {} as ChartLegendProps;

    const { baseElement, getByTestId } = render(
      <ThemeProvider theme={theme}>
        <ChartLegend data-testid="chart-legend" {...fakeProps} />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const chartLegend = getByTestId('chart-legend');

    expect(chartLegend).toBeInTheDocument();
  });
});
