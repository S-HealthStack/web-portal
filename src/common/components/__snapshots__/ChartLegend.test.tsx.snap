// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartLegend [NEGATIVE] should render chart legend with empty props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
}

<body>
  <div>
    <div
      class="c0"
      data-testid="chart-legend"
      mode="left"
    />
  </div>
</body>
`;

exports[`ChartLegend [NEGATIVE] should render chart legend without props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
}

<body>
  <div>
    <div
      class="c0"
      data-testid="chart-legend"
      mode="left"
    />
  </div>
</body>
`;

exports[`ChartLegend test chart legend if canToggle=false 1`] = `
.c4 {
  display: inline-grid;
  gap: 0.625rem;
  grid-template-columns: 1.25rem 1fr;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c7 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  -webkit-transition: color 300ms cubic-bezier(0.45,0.05,0,1);
  transition: color 300ms cubic-bezier(0.45,0.05,0,1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

.c6 {
  width: 1.25rem;
  height: 2.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 {
  display: inline-block;
}

.c2 input {
  display: none;
}

.c2 .c5 > svg > :nth-child(1) {
  -webkit-transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
  transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
  fill: #4475E3;
}

.c2 .c5 > svg > :nth-child(2) {
  -webkit-transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
  transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
}

.c2:hover > .c3 {
  cursor: pointer;
}

.c9 {
  display: inline-block;
}

.c9 input {
  display: none;
}

.c9 .c5 > svg > :nth-child(1) {
  -webkit-transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
  transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
  fill: #1AD598;
}

.c9 .c5 > svg > :nth-child(2) {
  -webkit-transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
  transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
}

.c9:hover > .c3 {
  cursor: pointer;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
}

.c1 {
  margin-right: 0.625rem;
}

.c1 circle {
  opacity: 1;
}

.c8 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #2D2D2D;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  width: 100%;
  -webkit-line-clamp: 2;
  line-clamp: 2;
  -webkit-box-orient: vertical;
  white-space: pre-line;
}

<body>
  <div>
    <div
      class="c0"
      data-testid="chart-legend"
      mode="left"
    >
      <div
        class="c1"
        mode="left"
      >
        <div
          class="c2 "
          color="primary"
          kind="filled"
          readonly=""
        >
          <label
            class="c3 c4"
            for="3"
          >
            <div
              class="c5 c6"
            >
              <svg />
            </div>
            <div
              class="c7"
            >
              <div
                class="c8"
              >
                Male
              </div>
            </div>
          </label>
          <input
            checked=""
            data-testid="male"
            id="3"
            readonly=""
            type="checkbox"
          />
        </div>
      </div>
      <div
        class="c1"
        mode="left"
      >
        <div
          class="c9 "
          color="secondaryGreen"
          kind="filled"
          readonly=""
        >
          <label
            class="c3 c4"
            for="4"
          >
            <div
              class="c5 c6"
            >
              <svg />
            </div>
            <div
              class="c7"
            >
              <div
                class="c8"
              >
                Female
              </div>
            </div>
          </label>
          <input
            checked=""
            data-testid="female"
            id="4"
            readonly=""
            type="checkbox"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ChartLegend test chart legend render 1`] = `
.c4 {
  display: inline-grid;
  gap: 0.625rem;
  grid-template-columns: 1.25rem 1fr;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c7 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  -webkit-transition: color 300ms cubic-bezier(0.45,0.05,0,1);
  transition: color 300ms cubic-bezier(0.45,0.05,0,1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

.c6 {
  width: 1.25rem;
  height: 2.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 {
  display: inline-block;
}

.c2 input {
  display: none;
}

.c2 .c5 > svg > :nth-child(1) {
  -webkit-transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
  transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
  stroke: #4475E3;
}

.c2 .c5 > svg > :nth-child(2) {
  -webkit-transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
  transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
}

.c2:hover > .c3 {
  cursor: pointer;
}

.c2:hover .c5 svg > :first-child {
  fill: #ECF1FC;
  stroke: #2F65E0;
}

.c9 {
  display: inline-block;
}

.c9 input {
  display: none;
}

.c9 .c5 > svg > :nth-child(1) {
  -webkit-transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
  transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
  stroke: #1AD598;
}

.c9 .c5 > svg > :nth-child(2) {
  -webkit-transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
  transition: stroke-color 300ms cubic-bezier(0.45,0.05,0,1), fill 300ms cubic-bezier(0.45,0.05,0,1);
}

.c9:hover > .c3 {
  cursor: pointer;
}

.c9:hover .c5 svg > :first-child {
  fill: #E8FBF5;
  stroke: #16B37F;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
}

.c1 {
  margin-right: 0.625rem;
}

.c1 circle {
  opacity: 1;
}

.c8 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #2D2D2D;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  width: 100%;
  -webkit-line-clamp: 2;
  line-clamp: 2;
  -webkit-box-orient: vertical;
  white-space: pre-line;
}

<body>
  <div>
    <div
      class="c0"
      data-testid="chart-legend"
      mode="left"
    >
      <div
        class="c1"
        mode="left"
      >
        <div
          class="c2 "
          color="primary"
          kind="radio"
        >
          <label
            class="c3 c4"
            for="1"
          >
            <div
              class="c5 c6"
            >
              <svg />
            </div>
            <div
              class="c7"
            >
              <div
                class="c8"
              >
                Male
              </div>
            </div>
          </label>
          <input
            data-testid="male"
            id="1"
            type="checkbox"
          />
        </div>
      </div>
      <div
        class="c1"
        mode="left"
      >
        <div
          class="c9 "
          color="secondaryGreen"
          kind="radio"
        >
          <label
            class="c3 c4"
            for="2"
          >
            <div
              class="c5 c6"
            >
              <svg />
            </div>
            <div
              class="c7"
            >
              <div
                class="c8"
              >
                Female
              </div>
            </div>
          </label>
          <input
            data-testid="female"
            id="2"
            type="checkbox"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
